quarkchain/core.py

```diff
@@ -664,21 +665,34 @@ def get_hash(self):
 
 
 def calculate_merkle_root(item_list):
+    ''' Using ETH2.0 SSZ style hash tree
+    '''
     if len(item_list) == 0:
-        return bytes(32)
-
-    sha_tree = []
-    for item in item_list:
-        sha_tree.append(sha3_256(item.serialize()))
+        sha_tree = [bytes(32)]
+    else:
+        sha_tree = [sha3_256(item.serialize()) for item in item_list]
 
+    zbytes = bytes(32)
     while len(sha_tree) != 1:
-        next_sha_tree = []
-        for i in range(0, len(sha_tree) - 1, 2):
-            next_sha_tree.append(sha3_256(sha_tree[i] + sha_tree[i + 1]))
         if len(sha_tree) % 2 != 0:
-            next_sha_tree.append(sha3_256(sha_tree[-1] + sha_tree[-1]))
-        sha_tree = next_sha_tree
-    return sha_tree[0]
+            sha_tree.append(zbytes)
+        sha_tree = [sha3_256(sha_tree[i] + sha_tree[i + 1]) for i in range(0, len(sha_tree), 2)]
+        zbytes = sha3_256(zbytes + zbytes)
+    return sha3_256(sha_tree[0] + len(item_list).to_bytes(8, "big"))
+
+
+def calculate_merkle_root1(item_list):
```

Comment:
It is intention is to verify two, but calculate_merkle_root1 is actually the original version defined in https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md#signed-roots.

Reply:
which one should we stick to in production? should have similar performance

Reply:
looks `calculate_merkle_root1 ` should be in test

Reply:
calculate_merkle_root1 will consume extra memory space since it zpads the list to p2_round so I would suggest to use calcuate_merkle_root although there is still space to improve.

Reply:
agree, then let's move `calculate_merkle_root1` to the test so to avoid future confusion on which function to choose

Reply:
done.